httpclient5/src/main/java/org/apache/hc/client5/http/impl/routing/DistributedProxySelector.java

```diff
@@ -0,0 +1,183 @@
+ * connection failures according to its own logic.
+ *
+ * @since 5.3
+ */
+@Contract(threading = ThreadingBehavior.SAFE)
+public class DistributedProxySelector extends ProxySelector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DistributedProxySelector.class);
+
+    /**
+     * A list of {@link ProxySelector} instances to be used by the DistributedProxySelector
+     * for selecting proxies.
+     */
+    private final List<ProxySelector> selectors;
+
+    /**
+     * A {@link ThreadLocal} variable holding the current {@link ProxySelector} for each thread,
+     * ensuring thread safety when accessing the current {@link ProxySelector}.
+     */
+    private final ThreadLocal<ProxySelector> currentSelector;
+
+    /**
+     * An {@link AtomicInteger} representing the shared index across all threads for
+     * maintaining the current position in the list of ProxySelectors, ensuring
+     * proper distribution of {@link ProxySelector} usage.
+     */
+    private final AtomicInteger sharedIndex;
+
+
+    /**
+     * Constructs a DistributedProxySelector with the given list of {@link ProxySelector}.
+     * The constructor initializes the currentSelector as a {@link ThreadLocal}, and
+     * the sharedIndex as an {@link AtomicInteger}.
+     *
+     * @param selectors the list of ProxySelectors to use.
+     * @throws IllegalArgumentException if the list is null or empty.
+     */
+    public DistributedProxySelector(final List<ProxySelector> selectors) {
+        if (selectors == null || selectors.isEmpty()) {
+            throw new IllegalArgumentException("At least one ProxySelector is required");
+        }
+        this.selectors = new ArrayList<>(selectors);
+        this.currentSelector = new ThreadLocal<>();
+        this.sharedIndex = new AtomicInteger();
+    }
+
+    /**
+     * Selects a list of proxies for the given {@link URI} by delegating to the current
+     * {@link ProxySelector} or the next available {@link ProxySelector} in the list if the current
+     * one returns an empty proxy list. If an {@link Exception} occurs, it will be caught
+     * and ignored, and the next {@link ProxySelector} will be tried.
+     *
+     * @param uri the {@link URI} to select a proxy for.
+     * @return a list of proxies for the given {@link URI}.
+     */
+    @Override
+    public List<Proxy> select(final URI uri) {
+        List<Proxy> result = Collections.emptyList();
+        ProxySelector selector = currentSelector.get();
+
+        for (int i = 0; i < selectors.size(); i++) {
+            if (selector == null) {
+                selector = nextSelector();
+                currentSelector.set(selector);
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Current selector is null, selecting next proxy selector for URI {}: {}", uri, selector);
+                }
+            }
+            try {
+                result = selector.select(uri);
```

Comment:
@arturobernalg Almost there. Please store the current selector only for the duration of the `result = selector.select(uri)` call and unset it immediately after (or in case of a failure in `#connectFailed`). The main idea is to use thread local only as short as possible and unset it as soon as possible to avoid thread local leaks.

Reply:
done